src/flag_gems/ops/dropout.py

```diff
@@ -4,83 +4,104 @@
 import triton
 import triton.language as tl
 
-from ..utils import libentry
 from ..utils.random_utils import philox_cuda_seed_offset
 
 try:
-    tl_rand_dtype = tl.int64
-
+    uint_to_uniform_float = tl.uint_to_uniform_float
+except AttributeError:
+    # Copied from triton.language package for compatibility
     @triton.jit
-    def _rand(seed, offset):
-        offset = offset.to(tl_rand_dtype)
-
-    _grid = (1,)
-    _seed, _offset = philox_cuda_seed_offset(0)
-    _rand[_grid](_seed, _offset)
-except Exception:
-    tl_rand_dtype = tl.int32
-
-del _grid
-del _seed
-del _offset
-
-
-@libentry()
-@triton.autotune(
-    configs=[
-        triton.Config({"N_BLOCK_SIZE": 256}, num_warps=2, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 256}, num_warps=2, num_stages=5),
-        triton.Config({"N_BLOCK_SIZE": 512}, num_warps=2, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 512}, num_warps=2, num_stages=5),
-        triton.Config({"N_BLOCK_SIZE": 1024}, num_warps=4, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 1024}, num_warps=4, num_stages=5),
-        triton.Config({"N_BLOCK_SIZE": 2048}, num_warps=4, num_stages=4),
-        triton.Config({"N_BLOCK_SIZE": 2048}, num_warps=4, num_stages=5),
-    ],
-    key=[
-        "N",
-    ],
+    def uint_to_uniform_float(x):
+        """
+        Numerically stable function to convert a random uint into a random float uniformly sampled in [0, 1).
+        """
+        # TODO: fix frontend issues and cleanup
+        # conditions can be simplified
+        # scale is ((2**23 - 1) / 2**23) * 2**(N_BITS - 1)
+        if tl.constexpr(x.dtype == tl.uint32) or tl.constexpr(x.dtype == tl.int32):
+            # maximum value such that `MAX_INT * scale < 1.0` (with float rounding)
+            x = x.to(tl.int32, bitcast=True)
+            scale = 4.6566127342e-10
+        else:
+            tl.static_assert(
+                tl.constexpr(x.dtype == tl.uint64) or tl.constexpr(x.dtype == tl.int64)
+            )
+            x = x.to(tl.int64, bitcast=True)
+            scale = 1.0842020432385337e-19
+        x = tl.where(x < 0, -x - 1, x)
+        return x * scale
+
+
+UNROLL = 4
+
+
+@triton.heuristics(
+    values={
+        # "BLOCK": lambda args: 512 if args["N"] <= 512 else 1024 if args["N"] <= 1024 else 2048,
+        "BLOCK": lambda args: 512 if args["N"] <= 512 else 1024,
+        "num_warps": lambda args: 4
+        if args["N"] <= 512
+        else 8
+        if args["N"] <= 1024
+        else 16,
+    }
 )
 @triton.jit
-def dropout_forward_kernel(
-    X,
-    Y,
-    N,
-    p,
-    philox_seed,
-    philox_offset,
-    N_BLOCK_SIZE: tl.constexpr,
-):
+def dropout_forward_kernel(X, Y, N, p, philox_seed, philox_offset, BLOCK: tl.constexpr):
```

Comment:
I suggest adding philox_seed, philox_offset into dot not specialize list to avoid specialization w.r.t them.

Reply:
Done.